MoreLinq/Return.cs

```diff
@@ -0,0 +1,106 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Mitch Bodmer. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+using System.Collections;
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Returns a single-element sequence containing the item provided.
+        /// </summary>
+        /// <typeparam name="T">The type of the item provided.</typeparam>
+        /// <param name="item">The item to wrap in a sequence.</param>
+        /// <returns>A sequence containing only the <paramref name="item"/>.</returns>
+
+        public static IEnumerable<T> Return<T>(T item) => new SingleElementList<T>(item);
+
+        internal sealed class SingleElementList<T> : IList<T>, IReadOnlyList<T>
+        {
+            private readonly T _item;
+
+            public SingleElementList(T item) => _item = item;
+
+            public IEnumerator<T> GetEnumerator() => new SingleElementEnumerator(_item);
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+            public void Add(T item) => throw new NotSupportedException("Single element list is immutable");
```

Comment:
Let's introduce:

```c#
Exception ReadOnlyError() =>
    throw new NotSupportedException("Single element list is immutable.");
```

Then have all other members call it, as in:

```c#
public void Add(T item) => throw ReadOnlyError();
```

Then there's one place to maintain the error and its message.

Reply:
Good call. Do you prefer that over:
```csharp
static NotSupportedException ReadOnlyException =>
    new NotSupportedException("Single element list is immutable.");

public void Add(T item) => throw ReadOnlyError;
```

I only suggest that because it might look a little cleaner in the call stack, and you still get the single place to change the error.

Reply:
> I only suggest that because it might look a little cleaner in the call stack

How does it look cleaner? Also, I don't believe that you can share a single exception instance as they are not immutable.

Reply:
I definitely wouldn't want to share a single exception. The property above should return a new exception every time since it's a lambda instead of an auto property.

I say it looks cleaner since the call stack would look like:
```
MyMethod
|--TheUnsupportedMethod
```
instead of
```
MyMethod
|--TheUnsupportedMethod
|----ReadOnlyError
```
since it's being thrown in the unsupported method and not the helper method.

Reply:
> I definitely wouldn't want to share a single exception. The property above should return a new exception every time since it's a lambda instead of an auto property.

Ah, a subtleness I just slipped by me since it is an unusual setup to have each invocation of a property produce a new instance of the error it represents.

> since it's being thrown in the unsupported method and not the helper method.

And now I realise my mistake. That `throw` should have never been there in `ReadOnlyError` since I had it at the call site. This is what I meant:

Exception ReadOnlyError() =>
    new NotSupportedException("Single element list is immutable.");

It's odd to see a property new-ing up an instance on each invocation of its getter so if we can stick with just regular method then we'll end up with the same goal of avoiding that last frame of stack.

Reply:
That makes more sense!

Yeah, I can see how that property can be deceiving. I'll go with the method.